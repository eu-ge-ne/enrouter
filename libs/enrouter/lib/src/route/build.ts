import * as regexparam from "regexparam";

import { logger } from "#lib/debug.js";

import type { Route } from "./mod.js";

const log = logger("route/build");

export interface BuildRoutesParams {
  modules: RouteModules;
}

/**
 * Builds `Route`s from `RouteModules`
 */
export function buildRoutes({ modules }: BuildRoutesParams): Route | undefined {
  log("Building routes");

  const entries = Object.entries(modules).sort(
    (a, b) => a[1].dirPath.length - b[1].dirPath.length,
  );

  const routes = new Map<string, Route>();

  function findParent(dp: string[]) {
    const parentPath = [...dp];

    while (parentPath.length > 0) {
      parentPath.pop();
      const path = parsePath("/" + parentPath.join("/"));
      const parent = routes.get(path);
      if (parent) {
        return parent;
      }
    }

    throw new Error("Parent not found");
  }

  for (const [moduleId, { dirPath, fileName, load }] of entries) {
    const isRoot = dirPath.length === 0;

    const path = isRoot ? "/" : parsePath("/" + dirPath.join("/"));

    let route = routes.get(path);
    if (!route) {
      route = {
        path,
        test: regexparam.parse(path, true),
        modules: [],
        loaded: false,
        elements: {},
      };
      routes.set(path, route);
    }

    if (!isRoot) {
      const parent = findParent(dirPath);
      if (!parent.tree?.find((x) => x.path === path)) {
        if (!parent.tree) {
          parent.tree = [];
        }
        parent.tree.push(route);
      }
    }

    route.modules.push({
      id: moduleId,
      fileName,
      load,
    });
  }

  const result = routes.get("/");

  log("Routes built: %o", routes);

  return result;
}

function parsePath(str: string) {
  return str.replace(/\[(.+)\]/, ":$1");
}

/**
 * Collection of route module descriptors.
 * Maps module id to fs path and async import function.
 */
type RouteModules = Record<
  string,
  {
    dirPath: string[];

    fileName: string;
    /**
     * Async module import function
     */
    load: () => Promise<unknown>;
  }
>;
