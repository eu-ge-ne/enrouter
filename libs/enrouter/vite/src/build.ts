import * as regexparam from "regexparam";

import { logger } from "#lib/debug.js";

import type { Route } from "#lib/route/mod.js";

const log = logger("route/build");

export type RouteModules = {
  id: string;
  dir: string[];
  fileName: string;
  load: () => Promise<unknown>;
}[];

/**
 * Builds `Route`s from `RouteModules`
 */
export function buildRoutes(modules: RouteModules): Route | undefined {
  log("Building routes");

  const sorted = modules.sort((a, b) => a.dir.length - b.dir.length);

  const routes = new Map<string, Route>();

  function findParent(dp: string[]) {
    const parentPath = [...dp];

    while (parentPath.length > 0) {
      parentPath.pop();
      const path = parsePath("/" + parentPath.join("/"));
      const parent = routes.get(path);
      if (parent) {
        return parent;
      }
    }

    throw new Error("Parent not found");
  }

  for (const { id, dir, fileName, load } of sorted) {
    const isRoot = dir.length === 0;

    const path = isRoot ? "/" : parsePath("/" + dir.join("/"));

    let route = routes.get(path);
    if (!route) {
      route = {
        path,
        test: regexparam.parse(path, true),
        modules: [],
        loaded: false,
        elements: {},
      };
      routes.set(path, route);
    }

    if (!isRoot) {
      const parent = findParent(dir);
      if (!parent.tree?.find((x) => x.path === path)) {
        if (!parent.tree) {
          parent.tree = [];
        }
        parent.tree.push(route);
      }
    }

    route.modules.push({
      id,
      fileName,
      load,
    });
  }

  const result = routes.get("/");

  log("Routes built: %o", routes);

  return result;
}

function parsePath(str: string) {
  return str.replace(/\[(.+)\]/, ":$1");
}

export function routeToJS(
  route: Route,
  getLoad: (id: string) => string,
  tab = 0,
): string {
  const tree = route.tree
    ? route.tree.map((x) => routeToJS(x, getLoad, 4))
    : undefined;
  const treeStr = tree
    ? `[
${tree.join(",\n")}
  ]`
    : undefined;

  const mods = route.modules
    .map(
      (x) => `
{
    id: "${x.id}",
    fileName: "${x.fileName}",
    load: ${getLoad(x.id)},
}`,
    )
    .join(",")
    .replace(/^/gm, " ".repeat(4));

  return `{
    path: "${route.path}",
    test: {
      keys: [${route.test.keys.map((x) => "${x}").join(",")}],
      pattern: ${route.test.pattern},
    },
    modules: [${mods}],
    loaded: false,
    elements: {},
    tree: ${treeStr},
}`.replace(/^/gm, " ".repeat(tab));
}
