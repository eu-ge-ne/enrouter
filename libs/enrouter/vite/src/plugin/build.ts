import * as regexparam from "regexparam";

import type { Route } from "#lib/route/mod.js";
import type { RouteModules } from "./modules.js";

/**
 * Builds `Route`s from `RouteModules`
 */
export function buildRoutes(modules: RouteModules): Route {
  const sorted = modules.sort((a, b) => a.dir.length - b.dir.length);

  const routes = new Map<string, Route>();

  function findParent(dp: string[]) {
    const parentPath = [...dp];

    while (parentPath.length > 0) {
      parentPath.pop();
      const path = parsePath("/" + parentPath.join("/"));
      const parent = routes.get(path);
      if (parent) {
        return parent;
      }
    }

    throw new Error("Parent not found");
  }

  for (const { dir, isRoot, id, fileName, importFn } of sorted) {
    const path = isRoot ? "/" : parsePath("/" + dir.join("/"));

    let route = routes.get(path);
    if (!route) {
      route = {
        path,
        test: regexparam.parse(path, true),
        modules: [],
        loaded: false,
        elements: {},
      };
      routes.set(path, route);
    }

    if (!isRoot) {
      const parent = findParent(dir);
      if (!parent.tree?.find((x) => x.path === path)) {
        if (!parent.tree) {
          parent.tree = [];
        }
        parent.tree.push(route);
      }
    }

    route.modules.push({
      id,
      fileName,
      importFn,
    });
  }

  const root = routes.get("/");
  if (!root) {
    throw new Error("Routes were not built");
  }

  return root;
}

function parsePath(str: string) {
  return str.replace(/\[(.+)\]/, ":$1");
}
