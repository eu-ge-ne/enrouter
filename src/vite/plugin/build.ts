import { parseRoutePath } from "./modules.js";
import type { Route } from "#lib/route/mod.js";
import type { RouteModules } from "./modules.js";

export function buildRouteTree(modules: RouteModules[]): Route {
  const routes = new Map<string, Route>();

  function findParent(dp: string[]): Route | undefined {
    if (dp.length === 0) {
      return;
    }

    const parentPath = [...dp];

    while (parentPath.length > 0) {
      parentPath.pop();
      const { path } = parseRoutePath(parentPath);
      const parent = routes.get(path);
      if (parent) {
        return parent;
      }
    }

    throw new Error("Parent not found");
  }

  const sorted = modules.sort((a, b) => a.dir.length - b.dir.length);

  for (const { dir, path, test, modules } of sorted) {
    let route = routes.get(path);
    if (!route) {
      route = {
        path,
        test,
        modules: modules.map(({ importStr, ...x }) => x),
        loaded: false,
        elements: {},
      };
      routes.set(path, route);
    }

    const parent = findParent(dir);
    if (parent) {
      parent.tree = parent.tree ?? [];
      parent.tree.push(route);
    }
  }

  const root = routes.get("/");
  if (!root) {
    throw new Error("Routes were not built");
  }

  return root;
}
