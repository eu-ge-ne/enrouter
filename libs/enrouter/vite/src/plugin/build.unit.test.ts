import { describe, test, expect } from "vitest";

import { buildRoutes } from "./build.js";

import type { RouteModules } from "./build.js";

describe("buildRoutes", () => {
  test("from 0 modules", () => {
    expect(() => buildRoutes([])).toThrowErrorMatchingSnapshot();
  });

  test("from 1 module", () => {
    const modules: RouteModules = [
      {
        id: "src/_layout.tsx",
        dir: [],
        fileName: "_layout.tsx",
        importFn: async () => undefined,
        importStr: '() => import("/home/src/_layout.tsx")',
      },
    ];

    expect(buildRoutes(modules)).toMatchSnapshot();
  });

  test("from 2 modules", () => {
    const modules: RouteModules = [
      {
        id: "src/_layout.tsx",
        dir: [],
        fileName: "_layout.tsx",
        importFn: async () => undefined,
        importStr: '() => import("/home/src/_layout.tsx")',
      },
      {
        id: "src/abc/_layout.tsx",
        dir: ["abc"],
        fileName: "_layout.tsx",
        importFn: async () => undefined,
        importStr: '() => import("/home/src/abc/_layout.tsx")',
      },
    ];

    expect(buildRoutes(modules)).toMatchSnapshot();
  });

  test("from 3 modules", () => {
    const modules: RouteModules = [
      {
        id: "src/_layout.tsx",
        dir: [],
        fileName: "_layout.tsx",
        importFn: async () => undefined,
        importStr: '() => import("/home/src/_layout.tsx")',
      },
      {
        id: "src/xyz/_layout.tsx",
        dir: ["xyz"],
        fileName: "_layout.tsx",
        importFn: async () => undefined,
        importStr: '() => import("/home/src/xyz/_layout.tsx")',
      },
      {
        id: "src/xyz/_index.tsx",
        dir: ["xyz"],
        fileName: "_index.tsx",
        importFn: async () => undefined,
        importStr: '() => import("/home/src/xyz/_layout.tsx")',
      },
    ];

    expect(buildRoutes(modules)).toMatchSnapshot();
  });
});
