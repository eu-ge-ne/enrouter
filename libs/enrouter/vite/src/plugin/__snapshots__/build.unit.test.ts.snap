// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildRoutes > from 0 modules 1`] = `[Error: Routes were not built]`;

exports[`buildRoutes > from 1 module 1`] = `
"{
    path: "/",
    test: {
      keys: [],
      pattern: /^(?=$|\\/)/i,
    },
    modules: [    
    {
        id: "src/_layout.tsx",
        fileName: "_layout.tsx",
        importFn: () => import("/home/src/_layout.tsx"),
    }],
    loaded: false,
    elements: {},
    tree: undefined,
}"
`;

exports[`buildRoutes > from 2 modules 1`] = `
"{
    path: "/",
    test: {
      keys: [],
      pattern: /^(?=$|\\/)/i,
    },
    modules: [    
    {
        id: "src/_layout.tsx",
        fileName: "_layout.tsx",
        importFn: () => import("/home/src/_layout.tsx"),
    }],
    loaded: false,
    elements: {},
    tree: [
    {
        path: "/abc",
        test: {
          keys: [],
          pattern: /^\\/abc(?=$|\\/)/i,
        },
        modules: [    
        {
            id: "src/abc/_layout.tsx",
            fileName: "_layout.tsx",
            importFn: () => import("/home/src/abc/_layout.tsx"),
        }],
        loaded: false,
        elements: {},
        tree: undefined,
    }
  ],
}"
`;

exports[`buildRoutes > from 3 modules 1`] = `
"{
    path: "/",
    test: {
      keys: [],
      pattern: /^(?=$|\\/)/i,
    },
    modules: [    
    {
        id: "src/_layout.tsx",
        fileName: "_layout.tsx",
        importFn: () => import("/home/src/_layout.tsx"),
    }],
    loaded: false,
    elements: {},
    tree: [
    {
        path: "/xyz",
        test: {
          keys: [],
          pattern: /^\\/xyz(?=$|\\/)/i,
        },
        modules: [    
        {
            id: "src/xyz/_layout.tsx",
            fileName: "_layout.tsx",
            importFn: () => import("/home/src/xyz/_layout.tsx"),
        },
        {
            id: "src/xyz/_index.tsx",
            fileName: "_index.tsx",
            importFn: () => import("/home/src/xyz/_layout.tsx"),
        }],
        loaded: false,
        elements: {},
        tree: undefined,
    }
  ],
}"
`;
